feat: add new LIP on coverage application

TheDZhon opened the pull request:
Specifies an in-protocol mechanism of applying coverage to address slashing/penalties events.
:heart_on_fire: :heart_on_fire: :heart_on_fire:

TheDZhon committed:
feat: add new LIP on coverage application

Specifies an in-protocol mechanism of applying coverage to address slashing/penalties events.

ujenjt commented on LIPS/LIP-draft_on_coverage_app.md:
````diff
@@ -0,0 +1,465 @@
+```
+So burning someone's shares (e.g. decreasing `totalShares` count) leads to increasing all of the other accounts' balances. 
+
+We propose to deploy a dedicated `LidoStETHBurner` contract which accepts burning requests by locking caller-provided stETH tokens. Those burning requests are initially set by the contract to a pending state. Actual burning happens within an oracle (`LidoOracle`) beacon report to prevent additional fluctuations of the existing stETH token rebase period.
````
Maybe rename to something more neutral like `CoverageApplicant`?

TheDZhon replied:
Yeeep, It's arguable naming questions :smile: 

Our proposal also supports non-coverage application cases, so explicitly using "Coverage*" may confuse and mislead someone, I think.

To suggest another one option, maybe ```LidoStEthValueRecoverer``` sounds good for you?

skozin replied:
I'd leave it as is since all this contract does is burning stETH, and it is not concerned about anything else. Higher-level mechanisms like the one that receives and applies cover should be built on top of it.

skozin replied:
Also, as per the proposal, we want to make this contract the only address allowed to burn stETH.

skozin replied:
I see how having smth that's called "stETH burner" in the protocol might frighten someone but I'd opt for transparency here. We can add a large comment on to the top of that contract explaining that it's not allowed to burn anyone else's stETH and is used for distributing value among stETH holders which is most easily done by burning some stETH.

skozin replied:
This mechanism of value distribution between token holders is well known among the Ethereum community since it's used not only by Ethereum itself (EIP-1559) but by other large protocols and even CEXes.

ujenjt commented on LIPS/LIP-draft_on_coverage_app.md:
```diff
@@ -0,0 +1,465 @@
+
+### Shares burnt counter
+
+We propose to store the total amount of shares ever burnt, distinguishing cover and non-cover shares, by maintaining two separate counters inside the `LidoStETHBurner` contract: `totalCoverSharesBurnt` and `totalNonCoverSharesBurnt`. The counters are increased when actual stETH burn is performed as part of the Lido Oracle report.
```
I like the idea to have a counter for `totalNonCoverSharesBurnt` but what's the exact case of using this counter?

TheDZhon replied:
I don't know the exact cases of using this counter for a particular reason too. :grin: 

But, in general, I think this way we preserve protocol transparency. Anyone can see the ever burnt shares amount (e.g. ```totalCoverSharesBurnt + totalNonCoverSharesBurnt```).

Also it could be used for stats and dashboards, for example, to explore `stETH`'s BME.

skozin replied:
Yeah, having the total number of shares ever burnt seems like a good idea in general since it might help in building future queries and dashboards. I'd put it the other way: given that implementing the total counter is trivial, what are the reasons for not doing this?

ujenjt commented on LIPS/LIP-draft_on_coverage_app.md:
```diff
@@ -0,0 +1,465 @@
+                                     uint256 _timeElapsed) external;
+}
+
+contract LidoStETHBurner is IBeaconReportReceiver {
```
Good one! I have several gas golfing optimizations in mind, and several code style considerations. Let's discuss on the coding stage

TheDZhon replied:
Ok, thx!

ujenjt commented on LIPS/LIP-draft_on_coverage_app.md:
```diff
@@ -0,0 +1,465 @@
+        throw;
+    }
+   
+    function recoverERC20(address token, uint256 amount) {
```
Good idea! Maybe we need something like this but for NFTs 😜

TheDZhon replied:
Nice catch! I am on it.

ujenjt reviewed (approved):
LGTM, please assign a number for LIP

TheDZhon commented:
@ujenjt and @skozin thank you so much! :handshake:

TheDZhon committed:
fix: assign lip id, apply PR review improvements

Assign lip number, go with LIP-6.
Rename `LidoStEthBurner` -> `SelfOwnedStEthBurner`.
Add recoverERC721.
Fix typos.
